packages/expo-file-system/android/src/main/java/expo/modules/filesystem/FileSystemModule.java

```diff
@@ -850,6 +947,68 @@ public void downloadResumablePauseAsync(final String uuid, final Promise promise
     }
   }
 
+  @ExpoMethod
+  public void askForDirectoryPermissions(final String initialFileUrl, Promise promise) {
+    if (mDirPermissionsRequest != null) {
+      promise.reject("ERR_FILESYSTEM_CANNOT_ASK_FOR_PERMISSIONS", "You have an unfinished permission request.");
+      return;
+    }
+
+    try {
+      Intent intent = new Intent(Intent.ACTION_OPEN_DOCUMENT_TREE);
+
+      if (Build.VERSION.SDK_INT >= Build.VERSION_CODES.O) {
+        Uri fileUri = initialFileUrl == null ? null : Uri.parse(initialFileUrl);
+        if (fileUri != null) {
+          intent.putExtra(DocumentsContract.EXTRA_INITIAL_URI, fileUri);
+        }
+      }
+
+      Activity activity = mModuleRegistry.getModule(ActivityProvider.class).getCurrentActivity();
+      if (activity == null) {
+        promise.reject("ERR_FILESYSTEM_CANNOT_ASK_FOR_PERMISSIONS", "Can't find activity.");
+        return;
+      }
+
+      mModuleRegistry.getModule(UIManager.class).registerActivityEventListener(this);
+
+      mDirPermissionsRequest = promise;
+      activity.startActivityForResult(intent, DIR_PERMISSIONS_REQUEST_CODE);
+
+    } catch (Exception e) {
+      Log.e(TAG, e.getMessage());
+      promise.reject("ERR_FILESYSTEM_CANNOT_ASK_FOR_PERMISSIONS", "Can't ask for permissions.", e);
+    }
+  }
+
+  @Override
+  public void onActivityResult(Activity activity, int requestCode, int resultCode, Intent data) {
+    if (requestCode == DIR_PERMISSIONS_REQUEST_CODE && mDirPermissionsRequest != null) {
+      if (resultCode == Activity.RESULT_OK) {
+        Uri treeUri = data.getData();
+
+        final int takeFlags = data.getFlags()
+          & (Intent.FLAG_GRANT_READ_URI_PERMISSION | Intent.FLAG_GRANT_WRITE_URI_PERMISSION);
+        activity.getContentResolver().takePersistableUriPermission(treeUri, takeFlags);
+
+        mDirPermissionsRequest.resolve(treeUri.toString());
+      } else {
+        mDirPermissionsRequest.resolve(false);
+      }
+
+      mModuleRegistry.getModule(UIManager.class).unregisterActivityEventListener(this);
+      mDirPermissionsRequest = null;
+    }
+  }
+
+  @Override
+  public void onNewIntent(Intent intent) {
+  }
+
+  private static boolean isSAFUri(Context context, Uri uri) {
```

Comment:
Maybe it would be good to add some explanation in the comment or be more clear by `isExternalStorageUri`?